### PR TITLE
Tweak addToExpression_reorder to avoid unary multiplication calls

### PR DESCRIPTION
**Pull request by joehuchette**

### CODE REVIEW

**Comment by joehuchette** [comment]
ref #537

File: src/v0.4/parseExpr_staged.jl [hunk 1]
```diff
@@ -305,32 +305,16 @@ end
...
-                push!(coef_expr.args, :(args[$i]))
-            end
-        end
-        return :(addToExpression(ex, $coef_expr, args[$var_idx]))
+    n = length(args)
+    varidx = find(t -> (t == Variable || t == AffExpr), collect(args))
+    allscalar = all(t -> (t <: Number), args[setdiff(1:n, varidx)])
+    if allscalar && length(varidx) == 1
+        idx = varidx[1]
+        @show :(addToExpression(ex, (*)(args[setdiff(1:$n,$idx)]...), args[$idx]))
```
Comment (mlubin) [inline src/v0.4/parseExpr_staged.jl]: Debug ``@show``?

File: src/v0.4/parseExpr_staged.jl [hunk 2]
```diff
@@ -305,32 +305,16 @@ end
...
+    elseif n == 2 # special-case to avoid calls to unary multiplication
+        :(addToExpression(ex, args[1], args[2]))
     else
-        #println("Nonlinear")
-        coef_expr = Expr(:call, :*, [:(args[$i]) for i in 1:(length(args)-1)]...)
-        return :(addToExpression(ex, $coef_expr, args[$(length(args))]))
+        :(addToExpression(ex, (*)(args[1:end-1]...), args[end]))
```
Comment (mlubin) [inline src/v0.4/parseExpr_staged.jl]: I think this form ends up splatting at runtime instead. That could be a big performance issue

File: src/v0.4/parseExpr_staged.jl [hunk 3]
```diff
@@ -305,32 +305,16 @@ end
 @generated addToExpression_reorder(ex, arg) = :(addToExpression(ex, 1.0, arg))
 
 @generated function addToExpression_reorder(ex, args...)
-    if !isleaftype(ex) || mapreduce(t -> !isleaftype(t), |, args)
-        error("Can't process abstract types")
-    end
-    # how many of the multiplicands are variables?
-    n_var = mapreduce(t -> (t == Variable || t == AffExpr), +, args)
-    has_quad = mapreduce(t -> (t == QuadExpr), |, args)
```
Comment (mlubin) [inline src/v0.4/parseExpr_staged.jl]: Can you figure out any reason why it was explicitly checking for quadratic terms? Maybe I had some idea to reorder to make quadratics efficient.

File: src/v0.4/parseExpr_staged.jl (hunk 2 above)
Comment (joehuchette) [inline src/v0.4/parseExpr_staged.jl]: How would you check something like that

File: src/v0.4/parseExpr_staged.jl [hunk 4]
```diff
@@ -305,32 +305,16 @@ end
...
-                push!(coef_expr.args, :(args[$i]))
-            end
-        end
-        return :(addToExpression(ex, $coef_expr, args[$var_idx]))
+    n = length(args)
+    varidx = find(t -> (t == Variable || t == AffExpr), collect(args))
+    allscalar = all(t -> (t <: Number), args[setdiff(1:n, varidx)])
+    if allscalar && length(varidx) == 1
+        idx = varidx[1]
+        :(addToExpression(ex, (*)(args[setdiff(1:$n,$idx)]...), args[$idx]))
```
Comment (mlubin) [inline src/v0.4/parseExpr_staged.jl]: Same for the splatting here.

**Comment by mlubin** [comment]
```
@generated function foo(args...)
    :( (*)(args...) )
end

function foo2(a,b,c)
    return a*b*c
end
```

```
julia> @code_llvm foo(1,2,3)

define %jl_value_t* @julia_foo_21346(%jl_value_t*, %jl_value_t**, i32) {
top:
  %3 = alloca [3 x %jl_value_t*], align 8
  %.sub = getelementptr inbounds [3 x %jl_value_t*]* %3, i64 0, i64 0
  %4 = getelementptr [3 x %jl_value_t*]* %3, i64 0, i64 2
  store %jl_value_t* inttoptr (i64 2 to %jl_value_t*), %jl_value_t** %.sub, align 8
  %5 = getelementptr [3 x %jl_value_t*]* %3, i64 0, i64 1
  %6 = load %jl_value_t*** @jl_pgcstack, align 8
  %.c = bitcast %jl_value_t** %6 to %jl_value_t*
  store %jl_value_t* %.c, %jl_value_t** %5, align 8
  store %jl_value_t** %.sub, %jl_value_t*** @jl_pgcstack, align 8
  store %jl_value_t* null, %jl_value_t** %4, align 8
  %7 = icmp eq i32 %2, 0
  br i1 %7, label %fail, label %pass

fail:                                             ; preds = %top
  call void @jl_bounds_error_tuple_int(%jl_value_t** %1, i64 0, i64 1)
  unreachable

pass:                                             ; preds = %top
  %8 = icmp ugt i32 %2, 1
  br i1 %8, label %pass2, label %fail1

fail1:                                            ; preds = %pass
  %9 = sext i32 %2 to i64
  call void @jl_bounds_error_tuple_int(%jl_value_t** %1, i64 %9, i64 2)
  unreachable

pass2:                                            ; preds = %pass
  %10 = icmp ugt i32 %2, 2
  br i1 %10, label %pass4, label %fail3

fail3:                                            ; preds = %pass2
  %11 = sext i32 %2 to i64
  call void @jl_bounds_error_tuple_int(%jl_value_t** %1, i64 %11, i64 3)
  unreachable

pass4:                                            ; preds = %pass2
  %12 = getelementptr %jl_value_t** %1, i64 1
  %13 = load %jl_value_t** %12, align 8
  %14 = load %jl_value_t** %1, align 8
  %15 = bitcast %jl_value_t* %13 to i64*
  %16 = bitcast %jl_value_t* %14 to i64*
  %17 = load i64* %15, align 16
  %18 = load i64* %16, align 16
  %19 = mul i64 %18, %17
  %20 = getelementptr %jl_value_t** %1, i64 2
  %21 = load %jl_value_t** %20, align 8
  %22 = bitcast %jl_value_t* %21 to i64*
  %23 = load i64* %22, align 16
  %24 = mul i64 %19, %23
  %25 = call %jl_value_t* @jl_box_int64(i64 signext %24)
  %26 = load %jl_value_t** %5, align 8
  %27 = getelementptr inbounds %jl_value_t* %26, i64 0, i32 0
  store %jl_value_t** %27, %jl_value_t*** @jl_pgcstack, align 8
  ret %jl_value_t* %25
}

julia> @code_llvm foo2(1,2,3)

define i64 @julia_foo2_21347(i64, i64, i64) {
top:
  %3 = mul i64 %1, %0
  %4 = mul i64 %3, %2
  ret i64 %4
}
```

**Comment by mlubin** [comment]
I'm not too sure that the old way was fast either:
```
@generated function foo3(args...)
    Expr(:call,:*,[:(args[$i]) for i in 1:length(args)]...)
end
```
```
julia> @code_llvm foo3(1,2,3)

define %jl_value_t* @julia_foo3_21174(%jl_value_t*, %jl_value_t**, i32) {
top:
  %3 = alloca [3 x %jl_value_t*], align 8
  %.sub = getelementptr inbounds [3 x %jl_value_t*]* %3, i64 0, i64 0
  %4 = getelementptr [3 x %jl_value_t*]* %3, i64 0, i64 2
  store %jl_value_t* inttoptr (i64 2 to %jl_value_t*), %jl_value_t** %.sub, align 8
  %5 = getelementptr [3 x %jl_value_t*]* %3, i64 0, i64 1
  %6 = load %jl_value_t*** @jl_pgcstack, align 8
  %.c = bitcast %jl_value_t** %6 to %jl_value_t*
  store %jl_value_t* %.c, %jl_value_t** %5, align 8
  store %jl_value_t** %.sub, %jl_value_t*** @jl_pgcstack, align 8
  store %jl_value_t* null, %jl_value_t** %4, align 8
  %7 = icmp eq i32 %2, 0
  br i1 %7, label %fail, label %pass

fail:                                             ; preds = %top
  call void @jl_bounds_error_tuple_int(%jl_value_t** %1, i64 0, i64 1)
  unreachable

pass:                                             ; preds = %top
  %8 = icmp ugt i32 %2, 1
  br i1 %8, label %pass2, label %fail1

fail1:                                            ; preds = %pass
  %9 = sext i32 %2 to i64
  call void @jl_bounds_error_tuple_int(%jl_value_t** %1, i64 %9, i64 2)
  unreachable

pass2:                                            ; preds = %pass
  %10 = icmp ugt i32 %2, 2
  br i1 %10, label %pass4, label %fail3

fail3:                                            ; preds = %pass2
  %11 = sext i32 %2 to i64
  call void @jl_bounds_error_tuple_int(%jl_value_t** %1, i64 %11, i64 3)
  unreachable

pass4:                                            ; preds = %pass2
  %12 = getelementptr %jl_value_t** %1, i64 1
  %13 = load %jl_value_t** %12, align 8
  %14 = load %jl_value_t** %1, align 8
  %15 = bitcast %jl_value_t* %13 to i64*
  %16 = bitcast %jl_value_t* %14 to i64*
  %17 = load i64* %15, align 16
  %18 = load i64* %16, align 16
  %19 = mul i64 %18, %17
  %20 = getelementptr %jl_value_t** %1, i64 2
  %21 = load %jl_value_t** %20, align 8
  %22 = bitcast %jl_value_t* %21 to i64*
  %23 = load i64* %22, align 16
  %24 = mul i64 %19, %23
  %25 = call %jl_value_t* @jl_box_int64(i64 signext %24)
  %26 = load %jl_value_t** %5, align 8
  %27 = getelementptr inbounds %jl_value_t* %26, i64 0, i32 0
  store %jl_value_t** %27, %jl_value_t*** @jl_pgcstack, align 8
  ret %jl_value_t* %25
}
```

File: src/v0.4/parseExpr_staged.jl [hunk 5]
```diff
@@ -304,34 +316,26 @@ end
 
 @generated addToExpression_reorder(ex, arg) = :(addToExpression(ex, 1.0, arg))
 
-@generated function addToExpression_reorder(ex, args...)
-    if !isleaftype(ex) || mapreduce(t -> !isleaftype(t), |, args)
-        error("Can't process abstract types")
+@generated function addToExpression_reorder(ex, x, y)
+    if x <: Union(Variable,AffExpr) && y <: Number
```
Comment (IainNZ) [inline src/v0.4/parseExpr_staged.jl]: Any reason to not use the generic equivalents?

File: src/v0.4/parseExpr_staged.jl [hunk 6]
```diff
@@ -304,34 +316,26 @@ end
...
-            end
-        end
-        return :(addToExpression(ex, $coef_expr, args[$var_idx]))
+    :(addToExpression(ex, x, y))
+end
+
+@generated function addToExpression_reorder(ex, args...)
+    n = length(args)
+    @assert n ≥ 3
+    varidx = find(t -> (t == Variable || t == AffExpr), collect(args))
```
Comment (IainNZ) [inline src/v0.4/parseExpr_staged.jl]: Same here

File: src/v0.4/parseExpr_staged.jl (hunk 5 above)
Comment (joehuchette) [inline src/v0.4/parseExpr_staged.jl]: We can't guarantee commutativity in some arbitrary type, so rewriting ``2*x*3`` to ``2*3*x`` might be valid. Though I guess we could just impose that restriction and use generics.

Comment (joehuchette) [inline src/v0.4/parseExpr_staged.jl]: We were pretty cavalier with breaking this for the longest time, which made the vectorized stuff more difficult than necessary, so I'm trying to be conservative here.

File: src/v0.4/parseExpr_staged.jl [hunk 7]
```diff
@@ -304,34 +316,26 @@ end
 
 @generated addToExpression_reorder(ex, arg) = :(addToExpression(ex, 1.0, arg))
 
-@generated function addToExpression_reorder(ex, args...)
-    if !isleaftype(ex) || mapreduce(t -> !isleaftype(t), |, args)
-        error("Can't process abstract types")
+@generated function addToExpression_reorder(ex, x, y)
+    if x <: Union(Variable,AffExpr) && y <: Number
+        x, y = y, x
```
Comment (mlubin) [inline src/v0.4/parseExpr_staged.jl]: Does this do anything below? ``x`` and ``y`` are just used as symbols in the expression returned

Comment (joehuchette) [inline src/v0.4/parseExpr_staged.jl]: Generated functions, man

**Comment by mlubin** [comment]
LGTM. Could you double check that this doesn't affect benchmark performance?

**Comment by joehuchette** [comment]
```
❯ j test/perf/macro.jl                                                       JuMP/git/tweak-staged !
  Running N=20...
    N=20 min 0.012497528
    N=20 min 0.026124674
  Running N=50...
    N=50 min 0.209613329
    N=50 min 0.462204239
  Running N=100...
    N=100 min 1.805433178
    N=100 min 4.228884544
```
```
❯ j test/perf/macro.jl                                                             JuMP/git/master !
  Running N=20...
    N=20 min 0.013070431
    N=20 min 0.025580498
  Running N=50...
    N=50 min 0.202727561
    N=50 min 0.496785758
  Running N=100...
    N=100 min 1.807605108
    N=100 min 4.051912484
```

**Comment by joehuchette** [comment]
So about a wash

**Comment by mlubin** [comment]
Seems fine

**Pull request merged**